File: calculator.cpp
#include <iostream>
#include <stack>
#include <math.h>

using namespace std;
class calculate
{
    stack<char> in;
    stack<int> post;

public:
    int priority(char formula)
    {
        if (formula == '^')
            return 3;
        else if (formula == '*' || formula == '/')
            return 2;
        else if (formula == '+' || formula == '-')
            return 1;
        else
            return 0;
    }
    double evaluate(string s)
    {
        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] >= '0' && s[i] <= '9')
            {
                post.push(s[i] - '0'); // subtracting from 0 char gives int
            }
            else
            {
                double operand2 = post.top();
                post.pop();
                double operand1 = post.top();
                post.pop();

                switch (s[i])
                {
                case '+':
                    post.push(operand1 + operand2);
                    break;
                case '-':
                    post.push(operand1 - operand2);
                    break;
                case '*':
                    post.push(operand1 * operand2);
                    break;
                case '/':
                    post.push(operand1 / operand2);
                    break;
                case '^':
                    post.push(pow(operand1, operand2));
                    break;

                default:
                    cout <<"Invalid symbole is being entered" << endl;
                    cout << "Enter the correct expression" << endl;
                    break;
                }
            }
        }
        return post.top();
    }

    void intopost(string s)
    {
        string converted;
        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(')
            {
                in.push(s[i]);
            }
            else if (s[i] == ')')
            {
                while (in.top() != '(' && !in.empty())
                {
                    converted += in.top();
                    in.pop();
                }
                in.pop();
            }
            else if (s[i] == '*' || s[i] == '/' || s[i] == '+' || s[i] == '-')
            {
                while (!in.empty() && priority(s[i]) <= priority(in.top()))
                {
                    converted += in.top();
                    in.pop();
                }
                in.push(s[i]);
            }
            else
                converted += s[i];
        }
        while (!in.empty()) {
            converted += in.top();
            in.pop();
        }
        cout << endl;
        cout << "The value of the above postfix expression is:-" << evaluate(converted) << endl;
    }
};

int main()
{
    calculate obj;
    cout << "Enter a expression to calculate it's value" << endl;
    string expression;
    getline(cin, expression);
    obj.intopost(expression);
    return 0;
}

File: Calculator.java
import java.util.Stack;

class Calculate {
    Stack<Character> in;
    Stack<Double> post;

    public Calculate() {
        in = new Stack<>();
        post = new Stack<>();
    }

    public int priority(char formula) {
        if (formula == '^')
            return 3;
        else if (formula == '*' || formula == '/')
            return 2;
        else if (formula == '+' || formula == '-')
            return 1;
        else
            return 0;
    }

    public double evaluate(String s) {
        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                post.push((double) (s.charAt(i) - '0')); // subtracting from 0 char gives int
            } else {
                double operand2 = post.pop();
                double operand1 = post.pop();

                switch (s.charAt(i)) {
                    case '+':
                        post.push(operand1 + operand2);
                        break;
                    case '-':
                        post.push(operand1 - operand2);
                        break;
                    case '*':
                        post.push(operand1 * operand2);
                        break;
                    case '/':
                        post.push(operand1 / operand2);
                        break;
                    case '^':
                        post.push(Math.pow(operand1, operand2));
                        break;
                    default:
                        System.out.println("Invalid symbol is being entered");
                        System.out.println("Enter the correct expression");
                        break;
                }
            }
        }
        return post.pop();
    }

    public void intopost(String s) {
        String converted = "";
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                in.push(s.charAt(i));
            } else if (s.charAt(i) == ')') {
                while (in.peek() != '(' && !in.empty()) {
                    converted += in.pop();
                }
                in.pop();
            } else if (s.charAt(i) == '*' || s.charAt(i) == '/' || s.charAt(i) == '+' || s.charAt(i) == '-') {
                while (!in.empty() && priority(s.charAt(i)) <= priority(in.peek())) {
                    converted += in.pop();
                }
                in.push(s.charAt(i));
            } else {
                converted += s.charAt(i);
            }
        }
        while (!in.empty()) {
            converted += in.pop();
        }
        System.out.println();
        System.out.println("The value of the above postfix expression is: " + evaluate(converted));
    }
}

public class Calculator {
    public static void main(String[] args) {
        Calculate obj = new Calculate();
        System.out.println("Enter an expression to calculate its value");
        String expression = System.console().readLine();
        obj.intopost(expression);
    }
}
